File: leetcode_java/java/com/goinhn/thorwcode/leetcode/question/_101_200/Sol121.java
package com.goinhn.thorwcode.leetcode.question._101_200;

/**
 * 121.买卖股票的最佳时机
 *
 * @author goinhn
 * @date 2021/11/23
 */
public class Sol121 {

    /**
     * time:(n)
     * space:(1)
     */
    private static class Solution1 {

        public int maxProfit(int[] prices) {
            int minPrice = Integer.MAX_VALUE;
            int maxGap = 0;
            for (int price : prices) {
                if (price < minPrice) {
                    minPrice = price;
                } else {
                    maxGap = Math.max(maxGap, price - minPrice);
                }
            }

            return maxGap;
        }

    }

}

File: leetcode_go/question/_101_200/_121/sol_121.go
// Package _121 买卖股票的最佳时间
// @author goinhn
// @date 2022/8/20
package _121

// 遍历
// time:O(n)
// space:O(1)
func maxProfit(prices []int) int {
	minPrice := prices[0]
	maxGap := 0
	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if maxGap < price-minPrice {
			maxGap = price - minPrice
		}
	}

	return maxGap
}
